### ISSUE TITLE
Problem with Greenlight talking to a Scalelite-Loadbalancer

### ISSUE THREAD

**Issue by bowman68** [bigbluebutton/greenlight#1264]
Hi folks,

i have problems bringing greenlight to work with a scalelite loadbalancer that connects to a couple of bbb-servers:

1. The Setup
- Greenlight is installed (without bbb-server and stuff) on a separate machine working as a frontend
- Scalelite is installed on a second machine as a loadbalancer
- A couple of BBB-Servers are installed on separate machines and are registered in Scalelite
- Our Learning Management System ILIAS (similar to moodle) is connected to Scalelite and allows joining meetings within courses

2. What works:
- The LMS ILIAS connected to Scalelite works fine: Lecturers can open Meetings, Students can join them, everything works as intended. Scalelite seems to work as intended.
- Greenlight configured to connect to a *dedicated BBB-Server* works as intended. But of course it should talk to Scalelite, not to the BBB-Servers directly

3. What doesn't work:
- Greenlight configured to talk to Scalelite only works half: You can sign in to Greenlight, edit your profile etc, but as soon as you switch to your homepage it says "Server Error, Bigbluebutton endpoint and secret invalid"

4. What i suspect:
- I think there's either a problem with the apis running on Scalelite or a problem with my setup of Greenlight:

5. How i installed Greenlight on the frontend-machine:
- installed a clean Nginx on the empty machine
- installed Docker and Docker-Compose on the empty machine
- installed and configured Greenlight as described in the manual
- checked it with

`docker run --rm --env-file .env bigbluebutton/greenlight:v2 bundle exec rake conf:check`

-> everything is fine
- configure the .env file with

```
BIGBLUEBUTTON_ENDPOINT=https://bbblb.kim.mydomain.edu/bigbluebutton/api
BIGBLUEBUTTON_SECRET=(...)
RELATIVE_URL_ROOT=/b
```

- created manually the path /etc/bigbluebutton/nginx/
- executed

`docker run --rm bigbluebutton/greenlight:v2 cat ./greenlight.nginx | sudo tee /etc/bigbluebutton/nginx/greenlight.nginx`

- copied the content of the vhost-file from a working standard-setup (BBB-Server with Greenlight on the same machine) in the current empty vhost-file on my Greenlight frontend, integrated the content of the (see above) /etc/bigbluebutton/nginx/greenlight.nginx and cut away some stuff created by BBB on the source-vhost. It now looks like this:

=======
```
server {
  server_name  bbb.mydomain.edu;
    listen   80;
    listen [::]:80;
    return 301 https://$server_name$request_uri; #redirect HTTP to HTTPS
  }

server {
  server_name bbb.mydomain.edu;
    listen 443 ssl;
    listen [::]:443 ssl;

    ssl_certificate /etc/nginx/ssl/bbb.mydomain.edu.crt;
    ssl_certificate_key /etc/nginx/ssl/bbb.mydomain.edu.key;
    ssl_session_cache shared:SSL:10m;
    ssl_session_timeout 10m;
    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers "ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5:!DSS:!AES256";
    ssl_prefer_server_ciphers on;
    ssl_dhparam /etc/nginx/ssl/dhp-4096.pem;

 location = / {
    return 307 /b;
  }

 access_log  /var/log/nginx/bigbluebutton.access.log;

location /b {
  proxy_pass          http://127.0.0.1:5000;
  proxy_set_header    Host              $host;
  proxy_set_header    X-Forwarded-For   $proxy_add_x_forwarded_for;
  proxy_set_header    X-Forwarded-Proto $scheme;
  proxy_http_version  1.1;
}

location /b/cable {
  proxy_pass          http://127.0.0.1:5000;
  proxy_set_header    Host              $host;
  proxy_set_header    X-Forwarded-For   $proxy_add_x_forwarded_for;
  proxy_set_header    X-Forwarded-Proto $scheme;
  proxy_set_header    Upgrade           $http_upgrade;
  proxy_set_header    Connection        "Upgrade";
  proxy_http_version  1.1;
  proxy_read_timeout  6h;
  proxy_send_timeout  6h;
  client_body_timeout 6h;
  send_timeout        6h;
}

}
```
=======

When checking the logs on Scalelite with

systemctl status scalelite-api.service scalelite-nginx.service

i see:

============

```
● scalelite-api.service - Scalelite API
   Loaded: loaded (/etc/systemd/system/scalelite-api.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2020-04-10 15:21:21 CEST; 1h 18min ago
  Process: 1755 ExecStartPre=/usr/bin/docker pull blindsidenetwks/scalelite:${SCALELITE_TAG}-api (code=exited, status=0/SUCCESS)
  Process: 1744 ExecStartPre=/usr/bin/docker rm scalelite-api (code=exited, status=0/SUCCESS)
  Process: 1190 ExecStartPre=/usr/bin/docker kill scalelite-api (code=exited, status=1/FAILURE)
 Main PID: 1776 (docker)
    Tasks: 15 (limit: 4915)
   CGroup: /system.slice/scalelite-api.service
           └─1776 /usr/bin/docker run --name scalelite-api --env-file /etc/default/scalelite --network scalelite --mount type=bind,source=/mnt/scalelite-recordings/var/bigbluebutton,target=/var/bigbluebutton blindsidenetwks/scalelite:v1-api

Apr 10 16:08:44 bbb_lb docker[1776]: I, [2020-04-10T14:08:44.736554 #128]  INFO -- : [7f212c21-da22-4dda-89f1-fc6ffbf2d920] Processing by BigBlueButtonApiController#get_recordings as XML
Apr 10 16:08:44 bbb_lb docker[1776]: I, [2020-04-10T14:08:44.736627 #128]  INFO -- : [7f212c21-da22-4dda-89f1-fc6ffbf2d920]   Parameters: {"meetingID"=>"e899d897597d091b448345f4ebff2d5140a27f54", "checksum"=>"e78807f652f1c40b4d9e6439ff5bde048a789f2c"}
Apr 10 16:08:44 bbb_lb docker[1776]: I, [2020-04-10T14:08:44.772626 #128]  INFO -- : [7f212c21-da22-4dda-89f1-fc6ffbf2d920] Completed 500 Internal Server Error in 36ms (ActiveRecord: 9.5ms | Allocations: 3016)
Apr 10 16:08:44 bbb_lb docker[1776]: F, [2020-04-10T14:08:44.773884 #128] FATAL -- : [7f212c21-da22-4dda-89f1-fc6ffbf2d920]
Apr 10 16:08:44 bbb_lb docker[1776]: [7f212c21-da22-4dda-89f1-fc6ffbf2d920] ActiveRecord::StatementInvalid (PG::UndefinedTable: ERROR:  relation "recordings" does not exist
Apr 10 16:08:44 bbb_lb docker[1776]: LINE 8:  WHERE a.attrelid = '"recordings"'::regclass
Apr 10 16:08:44 bbb_lb docker[1776]:                             ^
Apr 10 16:08:44 bbb_lb docker[1776]: ):
Apr 10 16:08:44 bbb_lb docker[1776]: [7f212c21-da22-4dda-89f1-fc6ffbf2d920]
Apr 10 16:08:44 bbb_lb docker[1776]: [7f212c21-da22-4dda-89f1-fc6ffbf2d920] app/controllers/bigbluebutton_api_controller.rb:247:in `get_recordings'

● scalelite-nginx.service - Scalelite Nginx
   Loaded: loaded (/etc/systemd/system/scalelite-nginx.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2020-04-10 15:21:22 CEST; 1h 18min ago
  Process: 1876 ExecStartPre=/usr/bin/docker pull blindsidenetwks/scalelite:${SCALELITE_TAG}-nginx (code=exited, status=0/SUCCESS)
  Process: 1839 ExecStartPre=/usr/bin/docker rm scalelite-nginx (code=exited, status=0/SUCCESS)
  Process: 1777 ExecStartPre=/usr/bin/docker kill scalelite-nginx (code=exited, status=1/FAILURE)
 Main PID: 2049 (docker)
    Tasks: 14 (limit: 4915)
   CGroup: /system.slice/scalelite-nginx.service
           └─2049 /usr/bin/docker run --name scalelite-nginx --env-file /etc/default/scalelite --network scalelite --publish 80:80 --publish 443:443 --mount type=bind,source=/mnt/scalelite-recordings/var/bigbluebutton/published,target=/var/bigbluebutton/published,readonly --mount type=bind,source=/etc/ssl/scalelite,target=/etc/nginx/ssl,readonly blindsidenetwks/scalelite:v1-nginx

Apr 10 15:23:05 bbb_lb docker[2049]: 134.34.33.18 - - [10/Apr/2020:13:23:05 +0000] "GET /bigbluebutton/api/create?name=ilias-bbb_ilias_uni_1157467_1&meetingID=ilias-bbb_ilias_uni_1157467_1&attendeePW=c5e6a5c6a5a3d6e5&moderatorPW=855e361a9dcdf5c8&voiceBridge=77746&logoutURL=https%3A%2F%2Filias-test.uni-konstanz.de%2Filias%2Fgoto.php%3Ftarget%3Dxbbb_887979%26client_id%3Dilias_uni&checksum=3be5f4bde6ef800a0859286936ca59c71ece71ec HTTP/1.1" 200 668 "-" "-" "-"
Apr 10 15:23:06 bbb_lb docker[2049]: 134.34.1.44 - - [10/Apr/2020:13:23:06 +0000] "GET /bigbluebutton/api/join?meetingID=ilias-bbb_ilias_uni_1157467_1&fullName=Sigmar+Papendick&password=855e361a9dcdf5c8&checksum=6323c50823c7ad048e9a3992990776eb8678cd6d HTTP/2.0" 302 276 "https://ilias-test.uni-konstanz.de/ilias/ilias.php?ref_id=887979&cmd=post&cmdClass=ilobjbigbluebuttongui&cmdNode=n9:j9&baseClass=ilObjPluginDispatchGUI&rtoken=8c58577b4d6e2c602f8dfbc524a4c3b3" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:75.0) Gecko/20100101 Firefox/75.0" "-"
Apr 10 15:23:36 bbb_lb docker[2049]: 134.34.33.18 - - [10/Apr/2020:13:23:36 +0000] "GET /bigbluebutton/api/getRecordings?meetingID=ilias-bbb_ilias_uni_1157467_1&checksum=f8376d0832acc7d55316e38a9e8e59d445e77ac4 HTTP/1.1" 500 131 "-" "-" "-"
Apr 10 15:23:36 bbb_lb docker[2049]: 134.34.33.18 - - [10/Apr/2020:13:23:36 +0000] "GET /bigbluebutton/api/getMeetingInfo?meetingID=ilias-bbb_ilias_uni_1157467_1&password=855e361a9dcdf5c8&checksum=40104971fb1ea566061b6cecfdf40f01d124bddb HTTP/1.1" 200 196 "-" "-" "-"
Apr 10 15:23:36 bbb_lb docker[2049]: 134.34.33.18 - - [10/Apr/2020:13:23:36 +0000] "GET /bigbluebutton/api/getMeetingInfo?meetingID=ilias-bbb_ilias_uni_1157467_1&password=855e361a9dcdf5c8&checksum=40104971fb1ea566061b6cecfdf40f01d124bddb HTTP/1.1" 200 196 "-" "-" "-"
Apr 10 15:23:37 bbb_lb docker[2049]: 134.34.33.18 - - [10/Apr/2020:13:23:37 +0000] "GET /bigbluebutton/api/getRecordings?meetingID=ilias-bbb_ilias_uni_1157467_1&checksum=f8376d0832acc7d55316e38a9e8e59d445e77ac4 HTTP/1.1" 500 131 "-" "-" "-"
Apr 10 15:23:37 bbb_lb docker[2049]: 134.34.33.18 - - [10/Apr/2020:13:23:37 +0000] "GET /bigbluebutton/api/getMeetingInfo?meetingID=ilias-bbb_ilias_uni_1157467_1&password=855e361a9dcdf5c8&checksum=40104971fb1ea566061b6cecfdf40f01d124bddb HTTP/1.1" 200 196 "-" "-" "-"
Apr 10 15:23:37 bbb_lb docker[2049]: 134.34.33.18 - - [10/Apr/2020:13:23:37 +0000] "GET /bigbluebutton/api/getMeetingInfo?meetingID=ilias-bbb_ilias_uni_1157467_1&password=855e361a9dcdf5c8&checksum=40104971fb1ea566061b6cecfdf40f01d124bddb HTTP/1.1" 200 196 "-" "-" "-"
Apr 10 16:08:44 bbb_lb docker[2049]: 134.34.241.36 - - [10/Apr/2020:14:08:44 +0000] "GET /bigbluebutton/api/isMeetingRunning?meetingID=e899d897597d091b448345f4ebff2d5140a27f54&checksum=10a8ebcd20cab9453e17b347cc45e0517cc12a5e HTTP/1.1" 200 118 "-" "Ruby" "-"
Apr 10 16:08:44 bbb_lb docker[2049]: 134.34.241.36 - - [10/Apr/2020:14:08:44 +0000] "GET /bigbluebutton/api/getRecordings?meetingID=e899d897597d091b448345f4ebff2d5140a27f54&checksum=e78807f652f1c40b4d9e6439ff5bde048a789f2c HTTP/1.1" 500 131 "-" "Ruby" "-"
~
~

```
===

The Greenlight-Logs say

==========
```
docker logs greenlight-v2

(...)
376ea086057b17875ee781.js.gz
autoprefixer: /usr/src/app/app/assets/stylesheets/application.scss:13187:3: Replace text-decoration-skip: ink to text-decoration-skip-ink: auto, because spec had been changed
I, [2020-04-10T13:21:58.908484 #30]  INFO -- : Writing /usr/src/app/public/assets/application-996333f5f97ef5e39521e0b7457a6340ba7f82ecf374b2943be363b2ca082a80.css
I, [2020-04-10T13:21:58.908811 #30]  INFO -- : Writing /usr/src/app/public/assets/application-996333f5f97ef5e39521e0b7457a6340ba7f82ecf374b2943be363b2ca082a80.css.gz
Puma starting in single mode...
* Version 3.12.4 (ruby 2.5.1-p57), codename: Llamas in Pajamas
* Min threads: 5, max threads: 5
* Environment: production
* Listening on tcp://0.0.0.0:80
Use Ctrl-C to stop
```
======

and

========
```
docker logs greenlight_db_1

(...)
PostgreSQL Database directory appears to contain a database; Skipping initialization

LOG:  database system was shut down at 2020-04-10 13:19:31 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  autovacuum launcher started
LOG:  database system is ready to accept connections
LOG:  invalid length of startup packet
ERROR:  database "greenlight_production" already exists
STATEMENT:  CREATE DATABASE "greenlight_production" ENCODING = 'utf8'
```
========

So basically i'm runnig out of ideas what to do next. There seems to be an issue with the Greelight -> Scalelite communication, is there something wrong with the containers running the apis (see above)? Or is there something missing in my crude vhost-setup on the Greenlight-Frontend?

Any help is appreciated, i've got to get this cluster working properly soon... :-/

Thanks,
Sigmar

**Comment by davidpesce** [comment]
@bowman68 - change your BIGBLUEBUTTON_ENDPOINT=https://bbblb.kim.mydomain.edu/bigbluebutton/ to include api/

It should look like:
```
BIGBLUEBUTTON_ENDPOINT=https://bbblb.kim.mydomain.edu/bigbluebutton/api
```

**Comment by bowman68** [comment]
> @bowman68 - change your BIGBLUEBUTTON_ENDPOINT=https://bbblb.kim.mydomain.edu/bigbluebutton/ to include api/
> 
> It should look like:
> 
> ```
> BIGBLUEBUTTON_ENDPOINT=https://bbblb.kim.mydomain.edu/bigbluebutton/api
> ```

...i've tried adding the /apt to the URL before (and configured it now again - updated my config above) but to no avail. I even copied the secret that works with our LMS ILIAS directly in the config to exclude typos - still no success.
I even thougt the length of the secret could be a problem, since the secret for Scalelite has 64 characters and the secets for the BBB-servers have 43. So i gave Scalelite a secret from one of our BBB-Servers and configured it in Greenlight, but that didn't help as well.

**Comment by bowman68** [comment]
I finaly could solve the problem: It was an issue with an inproper initialized Postgres-Database connected to Scalelite, that wasn't used by our LMS - so there the problem didn't appear.

**Comment by Furiouss38** [comment]
@bowman68 Could you explain what you did to solve your issue.
It seems I have the same, my scalelite with 3 bbb works in mooble, but my greenlight give me the same message you have

**Comment by bowman68** [comment]
…i’ve added the config-infos for Postgres in the thread.

**Comment by german2209** [comment]
> I finaly could solve the problem: It was an issue with an inproper initialized Postgres-Database connected to Scalelite, that wasn't used by our LMS - so there the problem didn't appear.
> 
> 1. I run Postgresql on the same machine as Scalelite. It's important to create a Postgres-User with the proper rights:
>    `createuser -P -l -s -e <database_username>`
> 2. In /etc/postgresql/10/main/pg_hba.conf add the permission to access the database:
> 
> ```
> hostssl all all <your machines IP/Network> md5
> 
> host all all <your machines IP/Network> md5
> 
> host all all 172.18.0.0/16 md5
> ```
> 
> The last line is the containers IP, not shure if it remains the same everywhere.
> 
> 1. Run the database-creation:
>    `docker exec -it scalelite-api bin/rake db:setup`
> 2. Restart Scalelite:
>    `systemctl restart scalelite.target`

Hi Bowman,
I'm experiencing the same troubles you described with Greenlight <--> Scalelite.
I followed step by step this guide: https://medium.com/@jesusfederico_39370/scalelite-lazy-deployment-745a7be849f6 that includes the docker exec -i scalelite-api bundle exec rake db:setup that when executed showed messages about creating tables and data, so I believe that Postgresql is runnig inside the docker container.
Did you install Postgresql outside of Docker? Can you share with me the steps and how did you configured your .env file to use the database?
Thanks in advance for your help!
Regards,
German Acevedo

**Comment by anmol1991** [comment]
Can Scalelite and Greenlite co-exit on a single machine ?

**Comment by basisbit** [comment]
@anmol1991 yes, you can run both on the same machine. Many of us do that.

**Comment by Mrityunjay6294** [comment]
Sir, I m trying to install greenlight by following the same instructions but it's not working for me
I followed this link https://medium.com/@jesusfederico_39370/scalelite-lazy-deployment-745a7be849f6 t

but I am not able to work with greenlight

**Comment by basisbit** [comment]
@Mrityunjay6294  For support requests, please use the [forum / mailing list](https://groups.google.com/forum/#!forum/bigbluebutton-greenlight). This issue tracker here is only for Greenlight development related issues. Sorry. Also, this issue here is already closed for almost a year.
